tiledb/filestore.py

```diff
@@ -0,0 +1,119 @@
+    :param tiledb.Ctx ctx: A TileDB context
+    """
+
+    def __init__(self, uri: str, ctx: "Ctx" = None) -> None:
+        self._ctx = ctx or default_ctx()
+        self._filestore_uri = uri
+
+    def write(self, buffer: ByteString, mime_type: str = "AUTODETECT") -> None:
+        """
+        Import data from an object that supports the buffer protocol to a Filestore Array.
+
+        :param buffer ByteString: Data of type bytes, bytearray, memoryview, etc.
+        :param str mime_type: MIME types are "AUTODETECT" (default), "image/tiff", "application/pdf"
+
+        """
+        try:
+            buffer = memoryview(buffer)
+        except TypeError:
+            raise TypeError(
+                f"Unexpected buffer type: buffer must support buffer protocol"
+            )
+
+        if not isinstance(mime_type, str):
+            raise TypeError(
+                f"Unexpected mime_type type '{type(mime_type)}': expected str"
+            )
+
+        lt.Filestore._buffer_import(
+            lt.Context(self._ctx.__capsule__(), False),
+            self._filestore_uri,
+            buffer,
+            lt.Filestore._mime_type_from_str(mime_type),
+        )
+
+    def read(self, offset: int = 0, size: int = -1) -> bytes:
+        """
+        :param int offset: Byte position to begin reading. Defaults to beginning of filestore.
+        :param int size: Total number of bytes to read. Defaults to -1 which reads the entire filestore.
+        :rtype: bytes
+        :return: Data from the Filestore Array
+
+        """
+        if not isinstance(offset, int):
+            raise TypeError(f"Unexpected offset type '{type(offset)}': expected int")
+
+        if not isinstance(size, int):
+            raise TypeError(f"Unexpected size type '{type(size)}': expected int")
+
+        if size == -1:
+            size = len(self)
+        size = max(size - offset, 0)
+
+        return lt.Filestore._buffer_export(
+            lt.Context(self._ctx.__capsule__(), False),
+            self._filestore_uri,
+            offset,
+            size,
+        )
+
+    def copy_from(self, uri: str, mime_type: str = "AUTODETECT") -> None:
+        """
+        Copy data from a file to a Filestore Array.
+
+        :param str uri: URI of file to export
+        :param str mime_type: MIME types are "AUTODETECT" (default), "image/tiff", "application/pdf"
+
+        """
+        if not isinstance(uri, str):
+            raise TypeError(f"Unexpected uri type '{type(uri)}': expected str")
+
+        if not isinstance(mime_type, str):
+            raise TypeError(
+                f"Unexpected mime_type type '{type(mime_type)}': expected str"
+            )
+
+        lt.Filestore._uri_import(
+            lt.Context(self._ctx.__capsule__(), False),
+            self._filestore_uri,
+            uri,
+            lt.Filestore._mime_type_from_str(mime_type),
+        )
+
+    def copy_to(self, uri: str) -> None:
```

Comment:
Not immediately necessary, but would be nice if `copy_to` checked if the output is a path and inferred the full URI based on the stored filename.